src/Compilers/CSharp/Portable/Lowering/Instrumentation/DynamicAnalysisInjector.cs

```diff
@@ -19,6 +19,7 @@ internal sealed class DynamicAnalysisInjector : CompoundInstrumenter
         private readonly MethodSymbol _method;
         private readonly BoundStatement _methodBody;
         private readonly MethodSymbol _createPayload;
+        private readonly PooledHashSet<DebugSourceDocument> _documentsBuilder;
```

Comment:
@cston Yes what you propose should be do-able. The only concern is that we would need to iterate over all the spans in each method (even in the common case where method only has spans in a single file) and large methods could potentially have several spans.

I think keeping around a pooled hashset object in a field is an acceptable trade-off to avoid the extra iteration. But if you feel strongly about this, I can get rid of the filed and do what you propose. Thoughts?